### PR TITLE
[VectorDistribution] Infer operand transpose for vector.contract distribution

### PR DESCRIPTION
**Pull request by Groverkss**
This patch also adds documentation on MFMA instruction layouts and a test for MTM contract type.

### CODE REVIEW

**Review by qedawkins — COMMENTED** [review]
Can we add a test?

**Review by harsh-nod — APPROVED** [review]
lgtm!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp [hunk 1]
```diff
@@ -18,15 +18,93 @@ using VectorValue = TypedValue<VectorType>;
...
+///
+/// Now that we have defined what layoutA, layoutB, layoutC are, we will define
+/// what the canonical layouts are for each MFMA instruction. These are
+/// represented as what part of the matrix each thread is carrying. These
+/// layouts were referenced from
+/// https://github.com/ROCm/amd_matrix_instruction_calculator
+///
+/// The naming scheme for these operators is InputType_MxNxK_OutputType.
 enum class MFMAType {
+  /// layoutA:
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:49]: Add a note that each number here represents the laneID responsible for that element.

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp [hunk 2]
```diff
@@ -18,15 +18,94 @@ using VectorValue = TypedValue<VectorType>;
...
+  /// ...
+  /// 15 15 15 15 31 31 31 31 47 47 47 47 63 63 63 63
+  ///
+  /// layoutB:
+  /// Transpose of layoutA
+  ///
+  /// layoutC:
+  /// Same as layoutB
   F16_16x16x16_F32,
+  /// layoutA:
```
Comment (harsh-nod) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:62]: This looks right. Also I visualized these a while back here: http://harshmenon.ai/n/codegen/wmma-and-mfma-instructions-on-amd-gpus/

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp [hunk 3]
```diff
@@ -18,15 +18,94 @@ using VectorValue = TypedValue<VectorType>;
...
+/// what the canonical layouts are for each MFMA instruction. These are
+/// represented as the original matrix, with elements representing which thread
+/// id in the subgroup gets which element.
+/// These layouts were referenced from
+/// https://github.com/ROCm/amd_matrix_instruction_calculator
+///
+/// The naming scheme for these operators is InputType_MxNxK_OutputType.
 enum class MFMAType {
+  /// layoutA:
+  ///  0  0  0  0 16 16 16 16 32 32 32 32 48 48 48 48
```
Comment (harsh-nod) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp:50]: this looks right as well.

**Pull request merged**